packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js

```diff
@@ -3887,6 +3924,132 @@ describe('ReactHooksWithNoopRenderer', () => {
       );
     });
 
+    it('unmount last state', () => {
+      function App(props) {
+        let A;
+        if (props.loadA) {
+          useState(0);
+        } else {
+          A = '[not loaded]';
+        }
+
+        return <Text text={`A: ${A}`} />;
+      }
+
+      ReactNoop.render(<App loadA={true} />);
+      expect(Scheduler).toFlushAndYield(['A: undefined']);
+      expect(ReactNoop.getChildren()).toEqual([span('A: undefined')]);
+      ReactNoop.render(<App loadA={false} />);
+
+      expect(() => {
+        // We don't throw because it would be noisy and require an additional prod check.
+        expect(Scheduler).not.toFlushAndThrow(
+          'Rendered fewer hooks than expected. This may be caused by an ' +
+            'accidental early return statement.',
+        );
+      }).toErrorDev([
```

Comment:
This is the warning the not pretty code mentioned above adds. As well as the "unmount effect" test below.